compiler/ilgen/VirtualMachineOperandArray.cpp

```diff
@@ -0,0 +1,183 @@
+ * (c) Copyright IBM Corp. 2017
+ *
+ *  This program and the accompanying materials are made available
+ *  under the terms of the Eclipse Public License v1.0 and
+ *  Apache License v2.0 which accompanies this distribution.
+ *
+ *      The Eclipse Public License is available at
+ *      http://www.eclipse.org/legal/epl-v10.html
+ *
+ *      The Apache License v2.0 is available at
+ *      http://www.opensource.org/licenses/apache2.0.php
+ *
+ * Contributors:
+ *    Multiple authors (IBM Corp.) - initial implementation and documentation
+ *******************************************************************************/
+
+#include "ilgen/VirtualMachineOperandArray.hpp"
+#include "compile/Compilation.hpp"
+#include "il/SymbolReference.hpp"
+#include "il/symbol/AutomaticSymbol.hpp"
+#include "ilgen/BytecodeBuilder.hpp"
+#include "ilgen/MethodBuilder.hpp"
+#include "ilgen/TypeDictionary.hpp"
+
+#define TraceEnabled    (TR::comp()->getOption(TR_TraceILGen))
+#define TraceIL(m, ...) {if (TraceEnabled) {traceMsg(TR::comp(), m, ##__VA_ARGS__);}}
+
+namespace OMR
+{
+VirtualMachineOperandArray::VirtualMachineOperandArray(TR::MethodBuilder *mb, int32_t numOfElements, TR::IlType *elementType, VirtualMachineRegister *arrayBaseRegister)
+   : VirtualMachineState(),
+   _mb(mb),
+   _numberOfElements(numOfElements),
+   _elementType(elementType),
+   _arrayBaseRegister(arrayBaseRegister)
+   {
+   int32_t numBytes = _numberOfElements * sizeof(TR::IlValue *);
+   _values = (TR::IlValue **) TR::comp()->trMemory()->allocateHeapMemory(numBytes);
+   memset(_values, 0, numBytes);
+
+   // store current operand stack pointer base address so we can use it whenever we need
+   // to recreate the stack as the interpreter would have
+   mb->Store("OperandArray_base", arrayBaseRegister->Load(mb));
+   }
+
+VirtualMachineOperandArray::VirtualMachineOperandArray(OMR::VirtualMachineOperandArray *other)
+   : VirtualMachineState(),
+   _mb(other->_mb),
+   _numberOfElements(other->_numberOfElements),
+   _elementType(other->_elementType),
+   _arrayBaseRegister(other->_arrayBaseRegister)
+   {
+   int32_t numBytes = _numberOfElements * sizeof(TR::IlValue *);
+   _values = (TR::IlValue **) TR::comp()->trMemory()->allocateHeapMemory(numBytes);
+   memcpy(_values, other->_values, numBytes);
+   }
+
+
+// commits the simulated operand array of values to the virtual machine state
+// the given builder object is where the operations to commit the state will be inserted
+// into the array which is assumed to be managed independently, most likely
+void
+VirtualMachineOperandArray::Commit(TR::IlBuilder *b)
+   {
+   TR::IlType *Element = _elementType;
+   TR::IlType *pElement = _mb->typeDictionary()->PointerTo(Element);
+
+   TR::IlValue *arrayBase = b->Load("OperandArray_base");
+
+   for (int32_t i = 0;i < _numberOfElements;i++)
+      {
+      TR::IlValue *element = _values[i];
+      if (element != NULL)
+         {
+         b->StoreAt(
+         b->       IndexAt(pElement,

```

Comment:
not standard indentation for compiler component files. I know one day we'll change it, but for now please keep it consistent even if more spaces makes the tree structure a tad more readable.